src/Buffer.mo

```diff
@@ -1,166 +1,1754 @@
+    let count = buffer.size();
+    if (length > count) {
+      Prim.trap "Buffer index out of bounds in prefix"
+    };
+
+    let newBuffer = Buffer<X>(newUpsize length); 
+
+    var i = 0;
+    while (i < length) {
+      newBuffer.add(buffer.get(i));
+      i += 1;
+    };
+
+    newBuffer
+  };
+
+  /// Checks if `prefix` is a prefix of `buffer`. Uses `equal` to
+  /// compare elements.
+  ///
+  /// Runtime: O(size of prefix)
+  ///
+  /// Space: O(size of prefix)
+  ///
+  /// *Runtime and space assumes that `equal` runs in O(1) time and space.
+  public func isPrefixOf<X>(prefix : Buffer<X>, buffer : Buffer<X>, equal : (X, X) -> Bool) : Bool {
+    let countPrefix = prefix.size();
+    if (buffer.size() < countPrefix) {
+      return false;
+    };
+
+    var i = 0;
+    while (i < countPrefix) {
+      if (not equal(buffer.get(i), prefix.get(i))) {
+        return false;
+      };
+      
+      i += 1;
+    };
+
+    return true;
+  };
+
+  /// Checks if `prefix` is a strict prefix of `buffer`. Uses `equal` to
+  /// compare elements.
+  ///
+  /// Runtime: O(size of prefix)
+  ///
+  /// Space: O(size of prefix)
+  ///
+  /// *Runtime and space assumes that `equal` runs in O(1) time and space.
+  public func isStrictPrefixOf<X>(prefix : Buffer<X>, buffer : Buffer<X>, equal : (X, X) -> Bool) : Bool {
+    let countPrefix = prefix.size();
+    if (buffer.size() <= countPrefix) {
+      return false;
+    };
+
+    var i = 0;
+    while (i < countPrefix) {
+      if (not equal(buffer.get(i), prefix.get(i))) {
+        return false;
+      };
+      
+      i += 1;
+    };
+
+    return true;
+  };
+
+  /// Returns the infix (sub-buffer) of `buffer` starting at index `start`
+  /// of length `length`. Traps if `start` is out of bounds, or `start + length`
+  /// is greater than the size of `buffer`.
+  ///
+  /// Runtime: O(length)
+  ///
+  /// Space: O(length)
+  public func infix<X>(buffer : Buffer<X>, start : Nat, length : Nat) : Buffer<X> {
+    let count = buffer.size();
+    let end = start + length; // exclusive
+    if (start >= count or end > count) {
+      Prim.trap "Buffer index out of bounds in infix"
+    };
+
+    let newBuffer = Buffer<X>(newUpsize length);
+
+    var i = start;
+    while (i < end) {
+      newBuffer.add(buffer.get(i));
+
+      i += 1;
+    };
+
+    newBuffer
+  };
+
+  /// Checks if `infix` is an infix of `buffer`. Uses `equal` to
+  /// compare elements.
+  ///
+  /// Runtime: O(size of infix + size of buffer)
+  ///
+  /// Space: O(size of infix)
+  ///
+  /// *Runtime and space assumes that `equal` runs in O(1) time and space.
+  public func isInfixOf<X>(infix : Buffer<X>, buffer : Buffer<X>, equal : (X, X) -> Bool) : Bool {
+    let infixSize = infix.size();
+
+    switch(indexOfBuffer(infix, buffer, equal)) {
+      case null infixSize == 0;
+      case _ true;
+    }
+  };
+
+  /// Checks if `infix` is a strict infix of `buffer`, i.e. `infix` must be
```

Comment:
Yes I agree. I actually have as a to do list item at the top of this PR to add examples to _all_ the functions, but @crusso suggested I do this in a separate PR since this one is large enough already.

Reply:
Infix is basically sub-buffer or sub-string.

I'll rename it to subbuffer. Just seemed natural to call it infix, since there's also a prefix and suffix, but I agree that it's not the most user friendly name.